crates/turborepo-lib/src/gitignore.rs

```diff
@@ -0,0 +1,141 @@
+#[cfg(unix)]
+use std::os::unix::fs::PermissionsExt;
+use std::{
+    fs::{File, OpenOptions},
+    io,
+    io::{BufRead, Write},
+};
+
+use turbopath::AbsoluteSystemPathBuf;
+
+const TURBO_GITIGNORE_ENTRY: &str = ".turbo";
+const GITIGNORE_FILE: &str = ".gitignore";
+
+pub fn ensure_turbo_is_gitignored(repo_root: &AbsoluteSystemPathBuf) -> Result<(), io::Error> {
+    let gitignore_path = repo_root.join_component(GITIGNORE_FILE);
+
+    if !gitignore_path.exists() {
+        let mut gitignore = File::create(gitignore_path)?;
+        #[cfg(unix)]
+        gitignore.metadata()?.permissions().set_mode(0o0644);
+        writeln!(gitignore, "{}", TURBO_GITIGNORE_ENTRY)?;
+    } else {
+        let gitignore = File::open(&gitignore_path)?;
+        let mut lines = io::BufReader::new(gitignore).lines();
+        let has_turbo =
+            lines.any(|line| line.map_or(false, |line| line.trim() == TURBO_GITIGNORE_ENTRY));
+        if !has_turbo {
+            let mut gitignore = OpenOptions::new()
+                .read(true)
+                .append(true)
+                .open(&gitignore_path)?;
```

Comment:
We have a wrapper for this with [`open_with_options`](https://github.com/vercel/turbo/blob/main/crates/turborepo-paths/src/absolute_system_path.rs#L422)